File: src/bin/day2.rs

use std::collections::HashMap;

fn count_char_occurences(string: &str) -> HashMap<char, usize> {
  let mut letters = HashMap::new();
  for ch in string.chars() {
    let counter = letters.entry(ch).or_insert(0);
    *counter += 1;
  }
  letters
}

fn count_repeat_chars (list: &Vec<String>, target_count: usize) -> usize {
  let mut count = 0;
  for id in list {
    let char_counts = count_char_occurences(&id);
    let counts: Vec<usize> = char_counts.values().map(|c| *c).collect();
    if counts.contains(&target_count) {
      count += 1;
    }
  }
  count
}

fn part_1 (input: &Vec<String>) {
  let count_2 = count_repeat_chars(input, 2);
  let count_3 = count_repeat_chars(input, 3);

  let chksum = count_2 * count_3;
  println!("Checksum = {}", chksum);
}

fn strings_are_close_enough (left: &String, right: &String) -> bool {
  let mut diffs = 0;

  for (ch_left, ch_right) in left.chars().zip(right.chars()) {
    if ch_left != ch_right {
      diffs += 1;
    }
  }

  diffs == 1
}

fn extract_common_chars (left: &String, right: &String) -> String {
  let mut common = String::with_capacity(left.len() - 1);

  for (ch_left, ch_right) in left.chars().zip(right.chars()) {
    if ch_left == ch_right {
      common.push(ch_left);
    }
  }

  common
}

fn part_2 (input: &Vec<String>) {
  for left in 0..input.len() {
    for right in left..input.len() {
      if left != right {
        let left_str = &input[left];
        let right_str = &input[right];
        if strings_are_close_enough(left_str, right_str) {
          let common_chars = extract_common_chars(left_str, right_str);
          println!("Found a match: ({}, {})", left_str, right_str);
          println!("Common characters: {}", common_chars);
        }
      }
    }
  }
}

fn main () {
  let input = include_str!("../data/day2.txt").lines().map(|l| l.to_owned()).collect();
  part_1(&input);
  part_2(&input);
}

File: day2.py
def count_char_frequency(box_id):
    chars = {}
    for char in box_id:
        chars.setdefault(char, 0)
        chars[char] += 1
    return chars


def count_repeating_chars(box_ids, target_frequency):
    def matches(box_id):
        return target_frequency in count_char_frequency(box_id).values()
    matches = filter(matches, box_ids)
    return len(list(matches))


def part_1(box_ids):
    two_chars = count_repeating_chars(box_ids, 2)
    three_chars = count_repeating_chars(box_ids, 3)
    chk = two_chars * three_chars
    print("Checksum = {}".format(chk))


def box_ids_are_close_enough(left, right):
    diffs = 0
    for (a, b) in zip(list(left), list(right)):
        if a is not b:
            diffs += 1
    return diffs is 1


def extract_common_chars(left, right):
    common = ""
    for (a, b) in zip(list(left), list(right)):
        if a is b:
            common += a
    return common


def part_2(box_ids):
    for left_ptr in range(0, len(box_ids)):
        for right_ptr in range(left_ptr + 1, len(box_ids)):
            left = box_ids[left_ptr]
            right = box_ids[right_ptr]
            if box_ids_are_close_enough(left, right):
                common = extract_common_chars(left, right)
                print("Found a match: ({}, {})".format(left, right))
                print("Common characters: {}".format(common))


def main():
    with open("day2.txt") as lines:
        box_ids = list(lines)
        part_1(box_ids)
        part_2(box_ids)


if __name__ == '__main__':
    main()
